src/serialize.h

```diff
@@ -59,6 +59,12 @@ inline T* NCONST_PTR(const T* val)
     return const_cast<T*>(val);
 }
 
+//! Safely convert odd char pointer types to standard ones.
+static inline char* CharCast(char* c) { return c; }
```

Comment:
nit: since these are inline, the static specifier is ignored

Reply:
Really?

My understanding was that a non-`static` `inline` function will still be emitted in non-inlined form in the object file, as the compiler can't know there are no non-inlinable calls from other translation units.

Making it static guarantees no calls from other translation units are possible, and thus permits the compliler to not emit any code as long as all calls within the current translation units are inlinable.

Reply:
That's my understanding too.

Reply:
It seems my understanding is correct for C, but not for modern C++. `inline` in C++ is not just a hint that the function is preferably inlined, but also automatically implies the function may violate ODR.

Thanks, this is very interesting.

Reply:
Fixed.

Reply:
> It seems my understanding is correct for C, but not for modern C++. inline in C++ is not just a hint that the function is preferably inlined, but also automatically implies the function may violate ODR.

Really confused. What is the reason for removing static and thinking it is ignored? What version of modern c++ is this assuming? I can't follow the logic here.

Reply:
Reading more on: http://en.cppreference.com/w/cpp/language/inline

`static` (or anonymous namespace) in combination with `inline` is not meaningless. It still indicates that different translation units may have different definitions.

However, non-`static` `inline` functions must be defined `inline` in every translation unit, and the compiler guarantees it will have the same address in all of them (if any). This means that the reason why I was using `static` here (being worried about having multiple emitted copies of a non-inlined version of the function) doesn't apply.

Reply:
> This means that the reason why I was using static here (being worried about having multiple emitted copies of a non-inlined version of the function) doesn't apply.

I'd still think you'd want to specify `inline` to tell the compiler that the function won't be accessed from a different translation unit and that exporting a symbol isn't necessary. It doesn't seem like it can be true that no symbol is emitted without `static` because according to your link, "An inline function with external linkage (e.g. not declared static)" "has the same address in every translation unit," which would require a symbol.

Testing locally, I see both `inline` and `static inline` functions emitting symbols, but `inline` functions have global symbols (`W` in nm output) while `static inline` functions only have local symbols (`t` in nm output).

Reply:
@ryanofsky That's what I'm seeing at `-O0` (when inlining is disabled). With `-O1` or higher, no symbol is emitted with either `inline` or `static inline` when all call sites are inlinable.

Also, when compiling with `g++`, with multiple object files where the call sites are not inlinable, each object gets indeed a symbol, but they're turned into a single symbol in the final executable. When compiling with `gcc`, the final executable has one symbol per original object.